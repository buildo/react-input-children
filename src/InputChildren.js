import React from 'react';
import { t, props } from 'tcomb-react';

export const Props = {
  children: t.maybe(t.ReactChildren),
  wrapper: t.maybe(t.Object)
};


/** `InputChildren` is a replacement for the base input react component capable of rendering
 * a child (link, button...) inside the input itself. It supports the same props of react input.
 * @param children - react children rendered inside the input
 * @param wrapper - props passed to wrapper 'div'
 */
@props(Props, { strict: false })
export default class InputChildren extends React.Component {

  static propTypes = {
    children: React.PropTypes.oneOfType([
      React.PropTypes.node,
      React.PropTypes.element
    ]),
    wrapper: React.PropTypes.object,
    innerRef: React.PropTypes.func
  }

  static defaultProps = {
    wrapper: {}
  }

  constructor(props) {
    super(props);
    this.state = { width: 1, height: 1 };
  }

  componentDidMount() {
    this.computeChildrenSize();
  }

  computeChildrenSize = () => {
    const childrenWrapper = this.refs.childrenWrapper.nodeType === 1 ?
      this.refs.childrenWrapper :
      this.refs.childrenWrapper.getDOMNode();
    const { clientWidth, clientHeight } = childrenWrapper;
    if (clientWidth !== this.state.width || clientHeight !== this.state.height) {
      this.setState({
        height: clientHeight,
        width: clientWidth
      });
    }
  }

  getInputStyle = () => {
    return {
      ...this.props.style,
      paddingRight: this.state.width,
      width: '100%',
      boxSizing: 'border-box'
    };
  }

  getChildrenStyle = () => {
    return {
      position: 'absolute',
      top: '50%',
      right: 0,
      marginTop: -1 * (this.state.height / 2)
    };
  }

  render() {
    const {
      children,
      wrapper,
      innerRef,
      ...inputProps
    } = this.props;

    const wrapperProps = {
      ...wrapper,
      style: {
        ...wrapper.style,
        position: 'relative'
      }
    };

    return (
      <div {...wrapperProps}>
        <input {...inputProps} style={this.getInputStyle()} ref={innerRef} />
        <div ref='childrenWrapper' style={this.getChildrenStyle()}>
          {children}
        </div>
      </div>
    );
  }

  componentDidUpdate() {
    this.computeChildrenSize();
  }

}
