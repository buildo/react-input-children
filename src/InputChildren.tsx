import * as React from 'react';

export namespace InputChildren {
  export type Props = React.HTMLProps<HTMLInputElement> & {
    /** React children rendered inside the input */
    children?: React.ReactNode,
    /** Props passed to wrapper 'div' */
    wrapper?:  React.HTMLProps<HTMLDivElement>
    /** Ref function for internal input reference */
    innerRef?: (input: HTMLInputElement) => void
  }
}

export type State = {
  width: number,
  height: number
}

/**
 * `InputChildren` is a replacement for the base input react component capable of rendering
 * a child (link, button...) inside the input itself. It supports the same props of react input.
 */
export class InputChildren extends React.Component<InputChildren.Props, State> {
  state = { width: 1, height: 1 }

  childrenWrapper: HTMLDivElement | null = null;
  computeTimeout: number | null = null

  componentDidMount() {
    this.computeChildrenSize();
  }

  computeChildrenSize = () : void => {
    if (this.childrenWrapper) {
      const { clientWidth, clientHeight } = this.childrenWrapper;
      if (clientWidth !== this.state.width || clientHeight !== this.state.height) {
        this.setState({
          height: clientHeight,
          width: clientWidth
        });
      }
    }
  }

  getInputStyle = (): React.CSSProperties => {
    return {
      ...this.props.style,
      paddingRight: this.state.width,
      width: '100%',
      boxSizing: 'border-box'
    };
  }

  getChildrenStyle = (): React.CSSProperties => {
    return {
      position: 'absolute',
      top: '50%',
      right: 0,
      marginTop: -1 * (this.state.height / 2)
    };
  }

  render() {
    const {
      children,
      wrapper = {},
      innerRef,
      ...inputProps
    } = this.props;

    const wrapperProps: React.HTMLProps<HTMLDivElement> = {
      ...wrapper,
      style: {
        ...wrapper.style,
        position: 'relative'
      }
    };

    return (
      <div {...wrapperProps}>
        <input {...inputProps} style={this.getInputStyle()} ref={innerRef} />
        <div ref={(ref: HTMLDivElement) => this.childrenWrapper = ref} style={this.getChildrenStyle()}>
          {children}
        </div>
      </div>
    );
  }

  componentDidUpdate() {
    this.computeTimeout = setTimeout(this.computeChildrenSize);
  }

  componentWillUnmount() {
    if (this.computeTimeout !== null) {
      clearTimeout(this.computeTimeout)
    }
  }
}
