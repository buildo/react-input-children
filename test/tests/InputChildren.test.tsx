import * as React from 'react';
import { render } from 'enzyme';
import InputChildren from '../../src';

describe('InputChildren', () => {

it('should render correctly', () => {
    const component = render(
      <div>
        <InputChildren className='input' wrapper={{ id: 'input-children' }}>>
          <a className='link'>Link</a>
        </InputChildren>
      </div>
    );
    expect(component.html()).toMatchSnapshot();
  });

//   it('input should have correct right padding', () => {
//     const component = mount(
//       <InputChildren className='input'>
//         <div style={{ width: 10 }} />
//       </InputChildren>
//     );

//     expect(component.find('input').instance().style['padding-right']).toBe('10px');
//   });

});
