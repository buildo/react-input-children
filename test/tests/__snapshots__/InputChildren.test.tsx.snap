// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputChildren should render correctly 1`] = `"<div id=\\"input-children\\" style=\\"position:relative\\"><input class=\\"input\\" style=\\"padding-right:1px;width:100%;box-sizing:border-box\\"><div style=\\"position:absolute;top:50%;right:0;margin-top:-0.5px\\">&gt;<a class=\\"link\\">Link</a></div></div>"`;

exports[`build build script generates every needed file 1`] = `
Array [
  "InputChildren.d.ts",
  "InputChildren.js",
  "index.d.ts",
  "index.js",
]
`;
